src/catalog/postgres/pg_proc_impl.cpp

```diff
@@ -0,0 +1,791 @@
+                                  namespace_oid_t procns, const std::string &procname,
+                                  const std::vector<type_oid_t> &arg_types) {
+  auto name_pri = procs_name_index_->GetProjectedRowInitializer();
+  byte *const buffer = common::AllocationUtil::AllocateAligned(pg_proc_all_cols_pri_.ProjectedRowSize());
+
+  auto name_pr = name_pri.InitializeRow(buffer);
+  auto name_map = procs_name_index_->GetKeyOidToOffsetMap();
+
+  auto name_varlen = storage::StorageUtil::CreateVarlen(procname);
+  auto all_arg_types_varlen = storage::StorageUtil::CreateVarlen(arg_types);
+  *reinterpret_cast<namespace_oid_t *>(name_pr->AccessForceNotNull(name_map[indexkeycol_oid_t(1)])) = procns;
+  *reinterpret_cast<storage::VarlenEntry *>(name_pr->AccessForceNotNull(name_map[indexkeycol_oid_t(2)])) = name_varlen;
+
+  std::vector<storage::TupleSlot> results;
+  procs_name_index_->ScanKey(*txn, *name_pr, &results);
+
+  proc_oid_t ret = INVALID_PROC_OID;
+  std::vector<proc_oid_t> matching_functions;
+  if (!results.empty()) {
+    const std::vector<type_oid_t> variadic = {dbc->GetTypeOidForType(type::TypeId::VARIADIC)};
+    auto variadic_varlen = storage::StorageUtil::CreateVarlen(variadic);
+
+    // Search through results and see if any match the parsed function by argument types
+    for (auto &tuple : results) {
+      auto table_pr = pg_proc_all_cols_pri_.InitializeRow(buffer);
+      bool UNUSED_ATTRIBUTE visible = procs_->Select(txn, tuple, table_pr);
+      storage::VarlenEntry index_all_arg_types = *reinterpret_cast<storage::VarlenEntry *>(
+          table_pr->AccessForceNotNull(pg_proc_all_cols_prm_[postgres::PgProc::PROALLARGTYPES_COL_OID]));
+      // variadic functions will match any argument types as long as there one or more arguments
+      if (index_all_arg_types == all_arg_types_varlen ||
+          (index_all_arg_types == variadic_varlen && !arg_types.empty())) {
+        proc_oid_t proc_oid = *reinterpret_cast<proc_oid_t *>(
+            table_pr->AccessForceNotNull(pg_proc_all_cols_prm_[postgres::PgProc::PROOID_COL_OID]));
+        matching_functions.push_back(proc_oid);
+        break;
+      }
+    }
+    if (variadic_varlen.NeedReclaim()) {
+      delete[] variadic_varlen.Content();
+    }
+  }
+
+  if (name_varlen.NeedReclaim()) {
+    delete[] name_varlen.Content();
+  }
+
+  if (all_arg_types_varlen.NeedReclaim()) {
+    delete[] all_arg_types_varlen.Content();
+  }
+
+  delete[] buffer;
+
+  if (matching_functions.size() == 1) {
+    ret = matching_functions[0];
+  } else if (matching_functions.size() > 1) {
+    // TODO(Joe Koshakow) would be nice to to include the parsed arg types of the function and the arg types that it
+    // matches with
+    throw BINDER_EXCEPTION(
+        fmt::format(
+            "Ambiguous function \"{}\", with given types. It matches multiple function signatures in the catalog",
+            procname),
+        common::ErrorCode::ERRCODE_DUPLICATE_FUNCTION);
+  }
```

Comment:
I'm not familiar enough with what we know and don't know at bind time, but this seems like an issue with allowing the creation of an illegal function overload rather than a valid bind-time failure.  Would the expanded index suggested in `builder.cpp` address this failure case?

Reply:
I think this should probably be considered a "bug".  I think adding the correct column (somehow) to the by-name index would enable enforcement of the underlying type constraint that this is catching too late.  I'm OK with this being marked resolved after an issue is created.

Reply:
#1361